File: viviane.py
from selenium import webdriver
from selenium.webdriver.firefox.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
import threading

# Configure Firefox options (disable headless mode to watch the video)
firefox_options = Options()
# firefox_options.headless = True  # Uncomment this to run in headless mode
firefox_options.add_argument("-private")  # Open in private mode

# Create a new Firefox profile and set preferences to allow autoplay
profile = webdriver.FirefoxProfile()
profile.set_preference("media.autoplay.default", 0)  # Allow all autoplay
profile.set_preference("media.autoplay.allow-muted", True)  # Allow muted autoplay

# Function to open a new browser window for each instance
def open_browser_instance(url):
    driver = webdriver.Firefox(firefox_profile=profile, options=firefox_options)  # Change this if using a different browser
    driver.get(url)
    
    # Wait for the video element to be present and click it to start playing
    try:
        video_element = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.CSS_SELECTOR, "video"))
        )
        driver.execute_script("arguments[0].scrollIntoView();", video_element)
        
        video_element.click()
    except Exception as e:
        print(f"Error clicking video: {e}")
    
    return driver

# URL of the YouTube video you want to watch
youtube_video_url = "https://www.youtube.com/watch?v=5EdVEghU-Xc"

# Function to open multiple browser instances simultaneously
def open_multiple_browsers(url, count):
    drivers = []
    threads = []
    for i in range(count):
        thread = threading.Thread(target=lambda i=i: drivers.append(open_browser_instance(url)))
        threads.append(thread)
        thread.start()
    for thread in threads:
        thread.join()
    return drivers

# Open 200 browser instances, 4 at a time
total_instances = 1
batch_size = 2
drivers = []

for i in range(0, total_instances, batch_size):
    print(f"Opening browsers {i+1} to {i+batch_size}")
    drivers.extend(open_multiple_browsers(youtube_video_url, batch_size))
    time.sleep(1)  # Add a delay to avoid overwhelming the server

# Perform tasks here if needed
# Example: Print the page title of each browser
for i, driver in enumerate(drivers):
    print(f"Browser {i+1} title: {driver.title}")

# Close all browsers
for driver in drivers:
    driver.quit()

File: viviane.js
const { Builder, By, until } = require('selenium-webdriver');
const firefox = require('selenium-webdriver/firefox');
const chrome = require('selenium-webdriver/chrome');
const { promisify } = require('util');
const sleep = promisify(setTimeout);

// Proxy list (you'll need a working list of proxy servers here)
const proxies = [
    "http://proxy1.com:8080",
    "http://proxy2.com:8080",
    "http://proxy3.com:8080"
];

// User-agent list for rotation
const userAgents = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.0 Mobile/15E148 Safari/604.1"
];

// Random sleep function for natural delays
function randomSleep(min, max) {
    const delay = Math.floor(Math.random() * (max - min + 1)) + min;
    return sleep(delay);
}

// Get a random proxy from the list
function getRandomProxy() {
    return proxies[Math.floor(Math.random() * proxies.length)];
}

// Get a random user agent
function getRandomUserAgent() {
    return userAgents[Math.floor(Math.random() * userAgents.length)];
}

// Random browsing actions for more human-like behavior
async function randomNavigation(driver) {
    const actions = driver.actions({ async: true });

    // Simulate scrolling to comments
    await actions.scroll(0, Math.floor(Math.random() * 600) + 200).perform();
    await randomSleep(3000, 7000); // Wait before the next action

    // Optionally click a random related video
    let relatedVideos = await driver.findElements(By.css('ytd-compact-video-renderer'));
    if (relatedVideos.length > 0) {
        let randomVideo = relatedVideos[Math.floor(Math.random() * relatedVideos.length)];
        await randomVideo.click();
        await randomSleep(5000, 10000);
    }

    // Return to the original video
    await driver.navigate().back();
}

// Open a browser instance with randomized proxy and user-agent
async function openBrowserInstanceWithRandomization(url, browser) {
    let driver;
    const proxy = getRandomProxy();
    const userAgent = getRandomUserAgent();

    if (browser === 'firefox') {
        const firefoxOptions = new firefox.Options()
            .addArguments('-private')
            .addArguments(`-user-agent=${userAgent}`);
        firefoxOptions.setProxy({ proxyType: 'manual', httpProxy: proxy, sslProxy: proxy });
        driver = await new Builder().forBrowser('firefox').setFirefoxOptions(firefoxOptions).build();
    } else if (browser === 'chrome') {
        const chromeOptions = new chrome.Options()
            .addArguments('--incognito')
            .addArguments(`--user-agent=${userAgent}`);
        chromeOptions.setProxy({ proxyType: 'manual', httpProxy: proxy, sslProxy: proxy });
        driver = await new Builder().forBrowser('chrome').setChromeOptions(chromeOptions).build();
    }

    await driver.get(url);

    // Wait for the video element to be present, then interact with it
    let videoElement = await driver.wait(until.elementLocated(By.css('video')), 10000);
    await randomSleep(3000, 6000); // Natural delay before starting video interaction
    await videoElement.click(); // Start playing the video

    // Perform random navigation for more human-like behavior
    await randomNavigation(driver);

    return driver;
}

// Open multiple browsers with randomization
async function openMultipleBrowsers(url, count, browser) {
    let drivers = [];
    for (let i = 0; i < count; i++) {
        const driver = await openBrowserInstanceWithRandomization(url, browser);
        drivers.push(driver);
        await randomSleep(5000, 10000); // Wait before opening next instance
    }
    return drivers;
}

// URL of the YouTube video
const youtubeVideoUrl = "https://www.youtube.com/watch?v=5EdVEghU-Xc&mute=1&autoplay=1";

// Main function to open browser instances
(async function() {
    const totalInstances = 4;
    let drivers = await openMultipleBrowsers(youtubeVideoUrl, totalInstances, 'firefox');

     // Wait for 45 seconds to simulate watching the video
     await sleep(45000); // Changed from 30000 to 45000

    // Print the page title of each browser
    for (let i = 0; i < drivers.length; i++) {
        let title = await drivers[i].getTitle();
        console.log(`Browser ${i + 1} title: ${title}`);
    }

    // Close all browsers after a set time
    for (let driver of drivers) {
        await driver.quit();
    }
})();
